lib/recog/nizer.rb

```diff
@@ -20,43 +20,122 @@ class Nizer
-    end
+    filter = { match_key: match_key, multi_match: false }
+    matches = self.match_all_db(match_string, filter)
+
+    return matches[0] if matches && matches[0]
+
     nil
   end
 
+  #
+  # @param match_key [String] Fingerprint DB name, e.g. 'smb.native_os'
+  # @param match_string [String] String to match
+  # @return [Array] Array of Fingerprint#match or empty array
   def self.multi_match(match_key, match_string)
-    match_string = match_string.to_s.unpack("C*").pack("C*")
-    @@db_manager ||= Recog::DBManager.new
+    filter = { match_key: match_key, multi_match: true }
+    matches = self.match_all_db(match_string, filter)
 
-    matches = Array.new #array to hold all fingerprint matches
+    matches
+  end
 
-    @@db_manager.databases.each do |db|
-      next unless db.match_key == match_key
+  #
+  # Search all fingerprint dbs and attempt to find matching
+  # {Fingerprint fingerprint}s. It will return the first match found
+  # unless the :multi_match option is used to request all matches.
+  # Returns an array of all matching fingerprints or an empty array.
+  #
+  # @param match_string [String] Service banner to match
+  # @param [Hash] filters This hash contains filters used to limit the
+  #   results to just those from specific types of fingerprints.
+  #   The values that these filters match come from the 'fingerprints' top
+  #   level element in the fingerprint DB XML or, in the case of 'protocol',
+  #   this value can be overridden at the individual fingerprint level by
+  #   setting a value for 'service.protocol'
+  #
+  #   With the exception of 'match_key', the filters below match the
+  #   'fingerprints' attributes with the same name.
+  # @option filters [String] :match_key Value from XML 'matches' or file name
+  # @option filters [String] :database_type fprint db type: service, util.os, etc.
+  # @option filters [String] :protocol Protocol (ftp, smtp, etc.)
+  # @option filters [String] :multi_match Return all matches instead of first
+  # @return [Array] Array of Fingerprint#match or empty array
+  def self.match_all_db(match_string, filters = {})
+    match_string = match_string.to_s.unpack('C*').pack('C*')
+    matches = Array.new # array to hold all fingerprint matches
 
+    self.load_db unless @@db_manager
+
+    return matches unless !@@db_manager.databases || !@@db_manager.databases.empty?
+
+    @@db_manager.databases.each do |db|
+      next if filters[:match_key] && !filters[:match_key].include?(db.match_key)
+      next if filters[:database_type] && !filters[:database_type].include?(db.database_type)
       db.fingerprints.each do |fp|
         m = fp.match(match_string)
-        matches.push(m) if m
+        if m
+          # Filter on protocol after match since each individual fp
+          # can contain its own 'protocol' value that overrides the
+          # one set at the DB level.
+          matches.push(m) unless filters[:protocol] && !filters[:protocol].include?(m['service.protocol'])
```

Comment:
Are you intentionally do partial string matching on `:protocol` here, or is the `:protocol` filter meant to be a list?  From what I can see, this is just a string that represents the desired protocol (if any), not a list.  Assuming that is true, I'd suggest either making this an exact match or updating to require a regular expression instead.  Otherwise, users will unknowingly try a filter of `http` and get `https` results (for example).